org.eclipse.dawnsci.analysis.dataset/src/org/eclipse/dawnsci/analysis/dataset/FunctionalUtils.java

```diff
@@ -268,6 +274,58 @@ public boolean tryAdvance(LongConsumer action) {
 			return gotNext;
 		}
 	}
+	
+	private static class LazyDatasetSpliterator extends Spliterators.AbstractSpliterator<ILazyDataset> {
+
+		private final SliceViewIterator it;
+		
+		public LazyDatasetSpliterator(SliceViewIterator it){
+			super(it.getTotal(),FLAGS);
+			this.it = it;
+		}
+
+		@Override
+		public boolean tryAdvance(Consumer<? super ILazyDataset> action) {
+			if (action == null) {
+				throw new NullPointerException();
+			}
+
+			boolean hasNext = it.hasNext();
+			if (hasNext) {
+				action.accept(it.next());
+			}
+
+			return hasNext;
+		}
+	}
+	
+	private static class DatasetSpliterator extends Spliterators.AbstractSpliterator<IDataset> {
+
+		private final SliceViewIterator it;
+		
+		public DatasetSpliterator(SliceViewIterator it){
+			super(it.getTotal(),FLAGS);
+			this.it = it;
+		}
+
+		@Override
+		public boolean tryAdvance(Consumer<? super IDataset> action) {
+			if (action == null) {
+				throw new NullPointerException();
+			}
+
+			boolean hasNext = it.hasNext();
+			if (hasNext) {
+				try {
+					action.accept(it.next().getSlice());
+				} catch (DatasetException e) {
+					throw new RuntimeException("Could not slice dataset",e);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS00112)